### ISSUE TITLE
remark-iframes: bad escape with allowDangerousHTML set to true

### ISSUE THREAD

**Issue by arobase-che** [zestedesavoir/zmarkdown#232]
Hi,

I got a problem each time i use iframes because i set the allowDangerousHTML option of remark-rehype to true.

This is my code : 

~~~js
const unified = require('unified')
const remarkParse = require('remark-parse')
const stringify = require('rehype-stringify')
const remark2rehype = require('remark-rehype')

const remarkIframe = require('remark-iframes')

const raw = require('rehype-raw')

unified()
  .use(remarkParse)
  .use(remarkIframe, {
    // this key corresponds to the hostname: !(http://hostname/foo)
    // the config associated to this hostname will apply to any iframe
    // with a matching hostname
    'www.youtube.com': {
      tag: 'iframe',
      width: 560,
      height: 315,
      disabled: false,
      replace: [
        ['watch?v=', 'embed/'],
        ['http://', 'https://'],
      ],
      thumbnail: {
        format: 'http://img.youtube.com/vi/{id}/0.jpg',
        id: '.+/(.+)$'
      },
      removeAfter: '&'
    }
  })
  .use(remark2rehype, {allowDangerousHTML: true}) 
  .use(raw)
  .use(stringify)
  .process( `!(https://www.youtube.com/watch?v=w0AOGeqOnFY)
<div>
Dangerous HTML 🐯
</div>`, (err, file) => {
    console.log(String(file));
  } );
~~~

As you can see, it's the example code with allowDangerousHTML activated (and so the plugin rehype-raw too) and processed to this text :

```raw
!(https://www.youtube.com/watch?v=w0AOGeqOnFY)
<div>
Dangerous HTML 🐯
</div>
```

The result is :  
```html
<iframe src="https://www.youtube.com/embed/w0AOGeqOnFY" width="560" height="315" allowfullscreen frameborder="0"></iframe>
&#x3C;div>
Dangerous HTML 🐯
&#x3C;/div>
```

As it should be : 
```html
<iframe src="https://www.youtube.com/embed/w0AOGeqOnFY" width="560" height="315" allowfullscreen frameborder="0"></iframe>
<div>
Dangerous HTML 🐯
</div>
```
Note : A dirty and quick fix can be to add a space before the element "!(" -> " !(". It seen to work.

:(

**Comment by arobase-che** [comment]
Ok, ok ...

I investigate a lot ... to out find a weird thing. It seem's to come from hast-util-from-parse5, i can see it from the HAST. 

```js
  console.log(inspect(fragment()));
  result = fromParse5(mode === FRAGMENT ? fragment() : document(), file);
  console.log(inspect(result));
```

in the package hast-util-raw. 

______

But a good fix to that issue is to put the tag name in upper case. "IFRAME" instead of "iframe". :upside_down_face: 
Then hast-util-from-parse5 parses it correctly ... Should we force it ?

**Comment by vhf** [comment]
Could you please provide a full example of how `hast-util-from-parse5` causes this issue and how `<IFRAME` would solve it?

**Comment by arobase-che** [comment]
:( Not really.

The only thing i can do is give you how i managed to investigate the problem.
First, it open the file index.js of the directory `node_modules/hast-util-raw`.
Then, from that file, i log everythings and observe that here : 
```js
  one.handlers.root = root;
  one.handlers.element = element;
  one.handlers.text = text;
  one.handlers.comment = comment;
  one.handlers.doctype = doctype;
  one.handlers.raw = raw;
  one.unknown = unknown;

  console.log(fragment());
  result = fromParse5(mode === FRAGMENT ? fragment() : document(), file);
  console.log(inspect(result));
```
Inside the object retourned by fragment, we can see that the html element is now a node text ... That obviously a bug.

So i tried to understand the fragment function. It's hard so i open the directory `parse5/lib/parser` of node_modules cause it was used in the fragment function.

And then finaly, you can see that they is a special case for the node named 'iframe' ... So you juste have to not use 'iframe' to not be in the special case and so it worked.

**Comment by arobase-che** [comment]
Ok, this issue is solved by rehypejs/rehype-raw@cca55675efe636625e71aa76f3f629786d2ef231 !


:smiley:
